src/provider/view/utils/Grouping.ts

```diff
@@ -0,0 +1,108 @@
+
+    /**
+     * The notifications to display within this group.
+     *
+     * This array should always contain at least one notification. If there are no
+     * notifications that fit within this category then the group should be deleted.
+     */
+    notifications: StoredNotification[];
+}
+
+/**
+ * When creating a new notification group, this function determines the
+ * user-visible title for the group
+ * @param notification Notificiation
+ * @param groupingType Grouping method
+ */
+export function getGroupTitle(notification: StoredNotification, groupingType: GroupingType): string {
+    if (groupingType === GroupingType.APPLICATION) {
+        return notification.source.name || notification.source.uuid;
+    } else if (groupingType === GroupingType.DATE) {
+        const {date} = notification.notification;
+        return moment(date).calendar(undefined, {
+            sameDay: '[Today]',
+            nextDay: '[Tomorrow]',
+            nextWeek: 'dddd',
+            lastDay: '[Yesterday]',
+            lastWeek: '[Last] dddd',
+            sameElse: 'DD/MM/YYYY'
+        });
+    }
+    throw new Error(`invalid groupMethod : ${groupingType}`);
+}
+
+/**
+ * Group notifications together based on the given grouping method e.g. application/date.
+ * @param notifications List of notifications to sort into groups.
+ * @param groupMethod Grouping type to use for sorting.
+ */
+export function groupNotifications(notifications: StoredNotification[], groupMethod: GroupingType): Map<string, Group> {
+    // TODO: This sorting should be removed and changed to use insert-sort inside the reduce below
+    // Pre-sort notifications by date (groups will then also be sorted by date)
+    notifications.sort((a: StoredNotification, b: StoredNotification) =>
+        b.notification.date.valueOf() - a.notification.date.valueOf());
+    // Reduce the notifications array into a Map of notifications grouped by their title
+    const groupMap = notifications.reduce((groups: Map<string, Group>, currentNotification: StoredNotification) => {
+        const groupTitle = getGroupTitle(currentNotification, groupMethod);
+        // If group title already exists just add it to the group
+        if (groups.has(groupTitle)) {
+            const group = groups.get(groupTitle)!;
+            groups.set(groupTitle, {
+                ...group,
+                notifications: [
+                    ...group.notifications,
+                    currentNotification
+                ]
+            });
+        } else {
+            // Create a new group and add the notification to the group
+            let key: string = currentNotification.source.name || currentNotification.source.uuid;
+            if (groupMethod === GroupingType.DATE) {
+                const date = new Date(currentNotification.notification.date);
+                key = [
+                    date.getUTCFullYear(),
+                    date.getMonth(),
+                    date.getDate()
```

Comment:
Probably shouldn't be mixing UTC and non-UTC functions here. Make the first one `date.getFullYear()`?